doc/api/esm.md

```diff
@@ -1130,18 +1130,40 @@ node --experimental-top-level-await b.mjs # works
 <!-- type=misc -->
 
 To customize the default module resolution, loader hooks can optionally be
-provided via a `--experimental-loader ./loader-name.mjs` argument to Node.js.
-
+provided via `--experimental-loader ./loader-name.mjs` arguments to Node.js.
 When hooks are used they only apply to ES module loading and not to any
 CommonJS modules loaded.
 
+Loaders are modules which export functions and data which can be used by
+Node.js to modify the behavior of loading ES Modules. Loaders may be “chained”
+together by passing multiple `--experimental-loader` arguments to Node.js. A
+loader module will be imported once as a singleton.
+
+Loaders follow an authority model, whereby each loader passed to node, from
+left to right, takes precedence over the next. For the `resolve`, `getFormat`,
+and `getSource` hooks, this means that the leftmost hook is called, and then
+the next, and so on, until one gives an answer. For `transformSource`, this
+means the rightmost hook is called first, and the result is passed to the next,
+and so on, until it reaches the leftmost hook. The `getGlobalPreloadCode` hooks
+run from left to right, so that the loaders to the left have more control over
+the environment than loaders to the right.
```

Comment:
currently if you want to modify the behaviour of something you just place it somewhere further right than yourself. that seems to be what you're asking?

Reply:
> currently if you want to modify the behaviour of something you just place it somewhere further right than yourself. that seems to be what you're asking?

What I would expect as a user is that the loaders run in the order they're defined, and that later-defined loaders have access to all the values returned by all the hooks of the earlier-defined loaders.

For example, if I'm the author of an instrumentation loader I would want to be able to have my loader go last and be able to see/override all return values of all previous loaders. I think the only way to do that is if all the hooks behave the same way, left-to-right.

Reply:
What do you think about this as a test we could add?

An “antivirus” loader, that’s loaded third after two other loaders. This antivirus loader will print a message to the console whenever it:

- Finds the string `virus` returned as any specifier from any other loader’s `resolve` hook.
- Sees any formats other than `module` or `commonjs` returned from any other loader’s `getFormat` hook.
- Finds the string `virus` within the source returned from any other loader’s `getSource`, `transformSource` or `getGlobalPreloadCode` hooks.

Reply:
> if I'm the author of an instrumentation loader I would want to be able to have my loader go last and be able to see/override all return values of all previous loaders. I think the only way to do that is if all the hooks behave the same way, left-to-right.

this is currently possible. just specify it as the first loader argument.

Reply:
I would think that the antivirus loader in my example should be the _last_ loader argument. After all, if you run:

```shell
node --require "./a.js" --require "./b.js"
```

Isn't `a.js` loaded before `b.js`? I would think that replacing `--require` with `--experimental-loader` should behave the same way.

After we flip that, do you mind showing me an example if how to implement this “antivirus” loader? Maybe add it as one of the tests? I think a test that shows that the last loader can override _anything_ returned by any previous loader would be the test that we need to proves that full instrumentation should be possible.

Reply:
> Isn't a.js loaded before b.js? I would think that replacing --require with --experimental-loader should behave the same way.

It does behave the same way. it might help to stop talking about execution order. this is about registration order. the further forward you register a loader, the more control it has.

```js
// virus loader

// Finds the string virus within any specifier returned from any other loader’s resolve hook.
export async function resolve(whatever, parent) {
  const result = await parent(whatever);
  if (result includes "virus") {
    throw new Error('dont load virus');
  }
  return result;
}

// Sees any formats other than builtin or module or commonjs returned from any other loader’s getFormat hook.
export async function getFormat(whatever, parent) {
  const result = await parent(whatever);
  if (result is not whitelisted type) {
  	throw new Error('dont load virus');
  }
  return result;
}

// Finds the string virus within the source returned from any other loader’s getSource, transformSource
export async function transformSource(source) {
  if (source includes "virus") {
    throw new Error('dont load virus');
  }
  return source;
}
```

this can't override or interact with other loader's global preload code because the security layering doesn't match up. the virus loader is validating that loaders don't load viruses, not that the loaders themselves aren't viruses.